Compile time check preflight returns no tec (RIPD-1624):

scottschurr opened the pull request:
Use compile-time techniques to make sure that the preflight portion of a transactor, the part that runs before the signature check, never returns a `tec` error code.

The changes are easiest to understand by starting with `TER.h`.  What was formerly the `TER` `enum` is divided up into six distinct `enum`s: `TELcodes`, `TEMcodes`, `TEFcodes`, `TERcodes`, `TEScodes`, and `TECcodes`.

Now that the sections of the error codes are subdivided, we need a way to controllably recombine them.  The `TERSubset` template class does that.  `TERSubset` uses a trait to determine which of the `TE*codes` `enum`s it can compare to and convert from.  `TERSubset` gets most of it lines of code from comparison operators.  This will (hopefully) get a lot better in C++20 with the `<=>` operator.

In TER.h there is one trait defined along with its specializations: `CanCvtToTER`.  We now redefine `TER`, which had been a simple `enum`, to this:
```
using TER = TERSubset<CanCvtToTER>;
```
This new definition of `TER` gives `TER` very similar characteristics to the old enum.

But for `preclaim()` we want to be able to choose a subset of codes that can be returned -- everything but `TECcodes`.  For this we make another trait to define a new kind of `TERSubset<>`: a kind that excludes `TECcodes`.  You see that in ApplySteps.h where we define the `CanCvtToNotTEC` trait.  Using that trait we define a new subset of `TE*codes` like this:
```
using NotTEC = TERSubset<CanCvtToNotTEC>;
```
We can now use `NotTEC` any places were all the codes other than `tec` are allowed.

But we still need to integrate `NotTEC` with `TER`.  So we specialize the `CanConvertToTER` trait one more time to allow a `NotTEC` to be compared with and converted into a `TER`.  That trait is also in ApplySteps.h, immediately following the definition of `NotTEC`:
```
// Allow NotTEC to be promoted to TER.
template <>
class CanCvtToTER<NotTEC> : public std::true_type
{ explicit CanCvtToTER() = delete; };
```
Now that we have `NotTEC` we can replace places that used to return `TER` codes with returning `NotTEC`.  If someone attempts to return a `TECcodes` member when they are returning a `NotTEC`, then they get a compile error.

There are a couple of places where the changes to the `TER` definition show up in the code and need to be adapted to:

- The biggest adaptation is that what used to be the `TER` `enum` is now six distinct `enum`s.  Usually that's papered over by our two conversion classes.  But when two different enums are used as arguments to the ternary operator, those arguments now need to be explicitly converted to the appropriate common type.  So you see changes like:
```
        return (flags & tapRETRY)
            ? TER {terNO_ACCOUNT}
            : TER {tecNO_ISSUER};
```
- The other adaptation is more subtle.  A non-class `enum` converts more easily into an integer than it converts into a `TERSubset`.  So in the `Status` class (in Status.h) I needed to do an `enable_if` to prevent a constructor from consuming `enums` as though they are integers:
```
   // The enable_if allows only integers (not enums).  Prevents enum narrowing.
   template <typename T,
       typename = std::enable_if_t<std::is_integral<T>::value>>
   Status (T code, Strings d = {}) ...
```

My questions for you:

1. Do you think this amount of machinery is appropriate to supply a compile time check.
2. If you do think the compile-time check is appropriate, are there ways to improve the machinery that I've coded up?

If we think this is worth keeping then I'll add some unit tests.  I want to make sure the team buys into the complexity of the technique before I sink more time into it.

The technique is somewhat open ended.  @ximinex and I talked briefly about constraining the transactors `doApply()` methods to only returning `tec` and `tes`.  That would be easy to do, or at least to validate, using the machinery given here.

Thanks for looking.

seelabs commented:
Enforcing constraints at compile time is a win. I'm a little concerned about the implementation, especially the constraint classes which "scatter" the code that decides what enums are allowed in the TERSubset. I.e. I want to look at the class `TER` and easily know that all enums are convertible without looking at trait classes scattered throughout the code.

I have three ideas to do this:

1) Use a variant. The idea is the class would contain a variant of like (pseudo code, I didn't code up any of this):
```
// All enums
using allEnums = boost::variant<TELcodes, TEMcodes, TEFcodes, TERcodes, TEScodes, TECcodes>;

// No tecs
using noTecs = boost::variant<TELcodes, TEMcodes, TEFcodes, TERcodes, TEScodes, TECcodes>;
```

Constraints are enforced when converting between the constrained types by using a conversion function like:
```
allEnums all;
noTecs notAll;
boost::apply_visitor([&all](auto&& v){all = v;}, notAll);
```

The above conversion will only work if `notAll` is a variant with a subset of types from `all` (or convertible to a type in `all`, but we don't have to worry about that). I.e. we can convert from `noTecs` to `allEnums` but not from `allEnums` to `noTecs`.

Of course, that `apply_visitor` will likely live in a copy constructor.

2) The second way is to assign a "category" to each of the enums that are powers of two, i.e.
```
constexpr int const telCategory = 1 << 0;
constexpr int const temCategory = 1 << 1;
constexpr int const tefCategory = 1 << 2;
// ect
```
Then have a class that takes the "allowable conversions" as a template parameter
```
template<int Categories>
class TER {
};
```
Then the constraints would be enforced by checking that the object being converted from has a `Categories` that is a subset of the
`TER` begin converted to (easily done with a bitwise and).

This will be a fast, easily understood solution. The one downside is the initial set of the constants would have to be defined differently than they are now, since they would need to know their category. I.e. something like a macro and:

```
deftef(FAILURE, -199);
deftef(ALREADY, -198);
...
deftec(CLAIM, 100);
deftec(PATH_PARTAIL, 101);
```

3) The third way is don't try to make this generic. Just define two classes, `allEnums` and `noTecs` and hard code the conversions needed. This actually isn't that bad a solution. It's easy for other devs to understand and it solves the immediate problem (however it is not easily extendable).

scottschurr commented:
Hi @seelabs, thanks for the review.  Before doing a major re-work, I'll put together a different branch that gathers the traits all in one place.  Having the trait distributed in different places seems like it is the primary issue?  We can see if localizing the traits is a sufficient improvement.

I was initially headed down the non-generic path.  And that might work really well if we had the `<=>` operator.  But the comparisons were turning into a mess, which is why I switched to using templates.

seelabs commented:
Let me name the three implementation cases "traits", "variant", and "bitsets" for the current implementation in the PR, proposal (1) above and proposal (2) above respectively.

Let's look at two use-cases: what does the declaration look like, and what does add a new subset class look like:

1) Declaration:

a) "traits"
Consider the declarations for "NotTEC" in each of the three
```
template <typename FROM>
class CanCvtToNotTEC : public std::false_type
{ explicit CanCvtToNotTEC() = delete; };

template <>
class CanCvtToNotTEC<TELcodes> : public std::true_type
{ explicit CanCvtToNotTEC() = delete; };

template <>
class CanCvtToNotTEC<TEMcodes> : public std::true_type
{ explicit CanCvtToNotTEC() = delete; };

template <>
class CanCvtToNotTEC<TEFcodes> : public std::true_type
{ explicit CanCvtToNotTEC() = delete; };

template <>
class CanCvtToNotTEC<TERcodes> : public std::true_type
{ explicit CanCvtToNotTEC() = delete; };

template <>
class CanCvtToNotTEC<TEScodes> : public std::true_type
{ explicit CanCvtToNotTEC() = delete; };

using NotTEC = TERSubset<CanCvtToNotTEC>;

// Allow NotTEC to be promoted to TER.
template <>
class CanCvtToTER<NotTEC> : public std::true_type
{ explicit CanCvtToTER() = delete; };
```

b) "variant"
```
using NotTEC = TERSubset<<TELcodes, TEMcodes, TEFcodes, TERcodes, TEScodes>
```

c) "bitset"
```
using NotTEC = TERSubset<<TELcodesID | TEMcodesID | TEFcodesID | TERcodesID | TEScodesID>
```

An alternate definition may look like (if we define an `AllTERID` or somesuch)
```
using NotTec = TERSubset<<AllTERID & ~TECcodesID>
```

Both the "variant" and "bitset" are easier to see what the subset is. In addition, we don't need a special trait to know that a `NotTEC` is convertible to a `TER`: The rule is "you can always convert a superset to a subset", and this should be easy to implement with both the "variant" and "bitset"
implementations.

2) Adding a new subset class:

a) "traits": Add a traits class with 5 specializations. In addition, we need to add a traits specialization for all the existing traits classes in order to convert subsets to supersets. (as the specialization of `CanCvtToTER<NotTEC>` does above.

b) "variant": Fill in the template with the subset

c) "bitset": Fill in the template with the bitset

Additional thoughts about of the "bitset" implementation: 
* We would have one type of class for these error codes (as opposed to the enums plus the constrained subset classes). 
* Of course, the cost of this is the declaration of the constants needs to be more complex (probably a macro). 
* Because it's one class, this "bitset" implementation should have an easy to implement comparison and equality operators.
* It's easy to write helper functions like `select` that can replace the ternary operators that return a type that's the union of the arguments.

scottschurr commented:
Hi @seelabs.  I'll suggest that you pick one of the alternatives you've identified and code it up.  That will improve the team's ability to compare options and choose a preferred approach.

As you approach the problem may be worth considering that the `tec` and `tes` codes are saved as the results of old transactions.  So changing the numeric values of these codes would require some kind of adaptation.

The other codes are not saved as transaction results.  I'm unsure whether these negative codes are ever returned by value through the API.  If they are, then we would need to think hard about changing their values as well.  If they are only returned as names, then we just need to preserve the name mapping, which should not be too hard.

seelabs commented:
@scottschurr I don't think it will be too hard to code up both the "variant" and "bitset" approaches. I'll do so. Neither approach will change the numeric values of the codes.

scottschurr commented:
Here's a slightly adjusted implementation that gathers all of the traits together to make them easier to see.  A minor concession to absolute correctness was made (I removed the explicit default constrictors of the traits) to improve readability.  https://github.com/ripple/rippled/compare/develop...scottschurr:ter-check-local-traits?expand=1

seelabs commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -47,7 +91,7 @@ struct PreflightResult
+    NotTEC const ter;
 
     /// Constructor
     template<class Context>
```
On line 104 here, we're initializing a variable of type `NotTEC` with a variable of type `TER`. This seems like a bug, I'd want the type system to catch this (TER could be a TEC, right?)

scottschurr replied:
Yup.  That looks like a bug.  I'd have expected a compile error out of that.  Investigating.

seelabs replied:
Too bad it's a PITA to write a unit test for a compile error :-) It's easy to see how this could slip through.

scottschurr replied:
There's a reason I haven't written the unit tests yet 😀.  I'm poking at it right now.  I'll let you know what I find (assuming I figure it out).

seelabs replied:
Actually, we should be able to use something like `static_assert(!is_convertible` now that I think about it.

scottschurr replied:
Found it.  It's the
```
constexpr explicit TERSubset (int rhs) : code_ (rhs) { } // For TxMeta
```
Apparently `explicit` wasn't enough to keep that constructor from being called.  For now I'm changing it to a named static factory function so the compiler won't be so helpful.  Good catch!  Thanks.

seelabs commented:
I coded up the "bitset" implementation I had in mind, see this patch: https://github.com/seelabs/rippled/commit/575b1af3e1e245782a7cb18958836c9a2b842526

The key class is here: https://github.com/seelabs/rippled/blob/s-ter-check/src/ripple/protocol/TER.h#L49-L83 

It uses a bitset to enforce constraints. Each error category is assigned an id that is a power of 2 (see the `TERCategory` enum). All the constant ids know their numeric values and categories (done with a macro, which I don't love but I can't think of a better way. Example: `deftel(LOCAL_ERROR          , -399);`). Variables know what categories they are allowed to hold (for now "all" and "no tecs"). You can only assign a value from a variable that has a subset of the allowed values (i.e. you are allowed to assign a `no tec` var to a `all` var, but not the other way around).

The biggest advantages I see over using traits classes are: 
1) It's easy to look at a declaration and know the constraints
2) It's easy to add new constraints 
3) All the comparison operators went away (about 80 LOC in the "traits" class implementation).
4) It's easy to combine variables of different constraints (the new constraint of two combined vars is the union of their bitsets. I don't know how to do that with traits. This can be useful to make the ternary choices cleaner...maybe).

scottschurr commented:
Looking into the Visual Studio CI failures.

HowardHinnant commented on src/ripple/protocol/TER.h:
```diff
@@ -311,7 +311,12 @@ class TERSubset
-    constexpr explicit TERSubset (int rhs) : code_ (rhs) { } // For TxMeta
+
+private:
+    constexpr TERSubset (int rhs) : code_ (rhs) { } // For TxMeta
```
Is the comment on this line obsolete?  If not, what does it mean?

HowardHinnant commented on src/ripple/protocol/TER.h:
```diff
@@ -218,6 +292,135 @@ enum TER
+
+    // Conversion operators.
+    // Conversion to int is not explicit to allow streaming without casting.
+    constexpr operator int() const { return code_; }
```
I started wondering how big of a pain it would be to make this conversion `explicit`, and thus a little safer.  It wasn't that bad: https://github.com/HowardHinnant/rippled/commit/fcd55a85f32470e1f72599850504ef77f54255b8

scottschurr replied:
Thanks for poking at that, @HowardHinnant.  That's a great thing to think about.  And it's cool that making the conversion operator explicit had basically no negative consequences.  You took the middle road and I took two (slightly) more radical roads.

1. @seelabs had noticed that if `TERSubset` has a conversion to `int` operator, then none of the comparison machines is needed.  So I made a branch that removes all of the comparison operators from `TERSubset`.  It worked just fine.  You can see the results here: https://github.com/scottschurr/rippled/commits/ter-check-no-cmp

2. For contrast, I thought it would be interesting to see if we could make it hard to accidentally convert from a `TER` to an `int`.  It turns out that the code base relies on being able to make those conversions.  But I thought they'd be easier to see if we gave names to those conversions.  So I tried it.  That branch also worked just fine.  You can see those results here: https://github.com/scottschurr/rippled/commits/ter-check-no-int

HowardHinnant replied:
I think I favor either named or explicit conversion to `int`. The implicit conversion to `int`, while simplifying the conversions, offers more potential for an accidental conversion leading to a run-time error.

Long-term, I'm currently leaning towards explicit conversions over named conversions as in general it reduces what I call the "apparent API".  I.e. one doesn't have to recall/lookup the name of the named conversion to convert from `T` to `U`.  It's always `U u{t};`  However that's long term, not short term.  gcc-5.2 won't even accept that syntax when `U` is `int`.  It's a bug that they fix somewhere in 6.x.  In the mean-time we have to limp along with `auto u = static_cast<U>(t);`.

Short-term, I'm fine with either named or explicit.

scottschurr commented:
I haven't (yet) heard anyone say this is a bad idea.  So I'm thinking we're going forward.  Speak up if you disagree!

It also seems the explorations so far are helping to refine questions about the characteristics we want from the implementation.

1. Do we want our type to easily interoperate with `int`?  Or do we want interoperation with `int` to be more difficult to do and easy to find in the code base?

2. Do we prefer a traits-based implementation (with some `enum`s)?  Or do we prefer the suggestion by @seelabs that we use a non-type template parameter to distinguish between error subsets (and lose the `enum`s)?

I don't think there are any clear technical considerations on either of these points.  As far as I can tell it's a matter of personal preference.

At this point, if I were left to my own devices, I'd probably implement using traits and do interoperation with `int`using named methods.  But I'm curious how other folks feel.  Remember, you may have to maintain this code yourself at some point in the future.  😀

seelabs commented:
Just to clarify, I think there are four implementation techniques on the table:
1) Traits, as in this PR
2) "bitsets", as implemented a commit linked in another comment
3) `std::variant`, not implemented, but briefly described in another comment
4) A non-generic hand coded implementations

I've outlined my arguments for my preferred approach (first "bitsets", then "std::variant", tho if we really wanted "std:variant" I should code it up, and so far no one has wanted it). In brief, I prefer the non-traits because:

1) It's trivial to add new constraints. To add the constraint @ximinex requested, we don't have to do anything, just declare a type of `TERIMP<tesCategory | tecCategory>` (or similar with the variant implementation), and it "just works" with all the correct conversions. With the traits we have to have to add a new traits class for each "leaf" enum class plus a new specialization for each conversion between the non-leaf classes.

2) The declarations are much more self-contained. Given the declaration above, it's easy to see what the constraint is just from the declaration, with traits we have to look at the traits specializations.

I also think that whatever implementation technique we use, the _concept_ of using the type-system to enforce these constraints is a good one and I'm in favor of merging that what ever implementation technique we decide on.

While the "bitset" implementation is my vote, I feel this is Scott S's baby and if he wants to use "traits", that's fine with me.

HowardHinnant commented:
I'm in favor of going forward on this.  On implementation technique, I'm open to anything, and partial to concrete code I can test and review. :-)

seelabs commented:
@HowardHinnant I did offer an implementation in this comment: https://github.com/ripple/rippled/pull/2490#issuecomment-383923335.

I coded up the "bitset" implementation I had in mind, see this patch: seelabs/rippled@575b1af

The key class is here: https://github.com/seelabs/rippled/blob/s-ter-check/src/ripple/protocol/TER.h#L49-L83

HowardHinnant commented:
Where is the "looking" emoji.... :-)

HowardHinnant commented:
The bitset implementation looks good to me too, except I have a firm preference for a non-implicit conversion to `int` (from `TERIMP`/`TERSubset`).  That could be accomplished with either design, as long as you throw back in the comparison operators.

seelabs commented:
@HowardHinnant OK, I concede. If you find both implementations equivalent then let's go with @scottschurr implementation in this PR.

seelabs commented on src/ripple/protocol/TER.h:
```diff
@@ -218,6 +292,140 @@ enum TER
+    operator Json::Value() const { return Json::Value {code_}; }
+
+    // Comparison
+    friend constexpr bool operator== (TERSubset lhs, TERSubset rhs)
```
`TERSubsets` with different constraints should still be comparable. Consider making the comparison functions templates, something like:
```
    template<template<typename> class T1, template<typename> class T2>
    friend constexpr bool operator== (TERSubset<T1> lhs, TERSubset<T2> rhs)
    { return lhs.code_ == rhs.code_; }
```

seelabs commented on src/ripple/protocol/TER.h:
```diff
@@ -218,6 +292,140 @@ enum TER
+    { return lhs.code_ == rhs.code_; }
+
+    friend constexpr bool operator!= (TERSubset lhs, TERSubset rhs)
+    { return ! (lhs == rhs); }
```
We need to write three implementations of the comparison functions for each comparison: one for `template<template<typename> class T1, template<typename> class T2>`, one for `TERSubset, enum type`, and one for `enum type, TERSubset`. It would be a win to get this down to one function. Maybe a `detail` scoped free function that converts both the enums and `TERSubset` to an int, do a `enable_if` based on if that free function is callable, then write the one comparison function.

scottschurr commented:
Thanks for the careful review @seelabs.  You make good points about a) comparability between `TERSubset`s and b) the volume of code to support comparisons.  I'll make a run at improving both of those points.

At the moment my brain is in RPC land for DepositPreauth, so I'll be a bit slow responding.  I'll try to have code to look at that (hopefully) addresses your concerns next week.

scottschurr commented:
- Rebase to -b5.
- Change `TER` -> `int` conversion to a named function (and include a comment justifying why a named conversion was chosen instead of an explicit conversion).
- Support comparison between all types that are convertible to `TER` (thanks @seelabs)
- Reduce redundancy in comparison operators (also thanks @seelabs).
- Add unit tests.

seelabs commented on src/ripple/protocol/TER.h:
```diff
@@ -218,6 +261,199 @@ enum TER
+
+    // Trait tells enable_if which types are allowed for construction.
+    template <typename T, typename = std::enable_if_t<Trait<T>::value>>
+    constexpr TERSubset (T rhs) : code_ ( TERtoInt (rhs)) { }
```
Nit: formatting

seelabs commented on src/test/protocol/TER_test.cpp:
```diff
@@ -49,13 +52,214 @@ struct TER_test : public beast::unit_test::suite
+            using To_t = std::decay_t<decltype (std::get<I1>(tup))>;
+            using From_t = std::decay_t<decltype (std::get<I2>(tup))>;
+            static_assert (!std::is_same<From_t, To_t>::value ^
+                std::is_convertible<From_t, To_t>::value, "Convert err");
```
`!a ^ b` seems like a strange way to write `a==b`. I feel like I'm missing something. Why the bit twiddling ops instead of `==`?

scottschurr replied:
Good catch.  Thanks.  The bit twiddling is because I was too deep into debugging the code to notice the equivalence.  `==` is much clearer.

seelabs commented on src/test/protocol/TER_test.cpp:
```diff
@@ -49,13 +52,214 @@ struct TER_test : public beast::unit_test::suite
+            static_assert (
+                ! std::is_constructible<To_t, int>::value, "Construct err");
+            static_assert (
+                ! std::is_assignable<To_t&, int const&>::value, "Assign err");
```
I think just `std::is_assignable<To_t&, int>`. Why the `int const&`?

scottschurr replied:
Mostly a reaction to uses of `is_assignable` being tricky to get right (as I learned the hard way), since it defaults types to `rvalue`s.  The From argument being a `const &` is the more typical case, since that's the signature of a defaulted copy assignment operator.  But, it's true, either would probably work.

seelabs commented on src/test/protocol/TER_test.cpp:
```diff
@@ -49,13 +52,214 @@ struct TER_test : public beast::unit_test::suite
+        }
+    };
+
+    // Fast iteration over the tuple.
```
I can see what this is doing: it's the Cartesian product for the types in the tuple. On the one hand, it's clever, but it's also not the clearest code. We need three separate cases to understand the code. I don't have an immediate solution (and I'm going to thumbs up), but I do think we should try to come up with something clearer in the future.

scottschurr replied:
I'm not much of a metaprogrammer.  This was the simplest way I could think of to handle the type iteration without explicitly listing each case, starting with a `tuple` as a way to make a type list.  I wouldn't be surprised if there are better ways to do it.  I'm open to suggestions.

seelabs commented on src/ripple/protocol/TER.h:
```diff
@@ -305,7 +305,9 @@ class TERSubset
     template <typename T, typename = std::enable_if_t<Trait<T>::value>>
-    constexpr TERSubset (T rhs) : code_ ( TERtoInt (rhs)) { }
+    constexpr TERSubset (T rhs)
+    : code_ ( TERtoInt (rhs))
```
Nit: There's still the extra space here. (I really don't care, but it's not how we normally format this).

scottschurr replied:
Just to be clear before I change anything, you're referring to the space between the paren and `TERtoInt`?

seelabs replied:
Yes. I think most of us are writing `code_(TERtoInt(rhs))` these days, tho I think the style guide officially says it should be `code_ (TerToInt (rhs))`. Either way if fine with me, but yeah, we should get rid of the space between the opening paren and parameter for sure.

seelabs commented:
Still :+1: after dd56e8d commit

seelabs commented:
In 1.1.0-b1
